File: evtol_simulation/trajectory/DifferentialFlatness.m
classdef DifferentialFlatness < handle
% DIFFERENTIALFLATNESS  Polynomial trajectory in flat outputs (x,y,z,psi).
%
% A vectored-thrust eVTOL is differentially flat in (sigma) = [x; y; z; psi].
% Once a smooth sigma(t) up to 4th derivative is given, the body attitude,
% thrust, and angular rates are recovered algebraically:
%
%   T (vec)  =  m * (sigma_ddot - g_NED)        (specific thrust in NED)
%   x_B      =  T / |T|
%   y_B      =  ... (heading constraint)
%   omega_B  =  recovered from third derivative (jerk)
%   omega_dot_B = recovered from fourth derivative (snap)
%
% Trajectory class: minimum-snap polynomials between waypoints
% (Mellinger & Kumar 2011). Each segment k between waypoints uses 8th-order
% polynomial in t to make all derivatives up to snap continuous.

    properties
        waypoints       % N x 4: [x y z psi] at each waypoint
        time_alloc      % (N-1) x 1 time per segment
        coeffs          % 4 x 8 x (N-1)  [x;y;z;psi] coefficients per segment
    end

    methods
        function obj = DifferentialFlatness(waypoints, time_alloc)
            obj.waypoints = waypoints;
            obj.time_alloc = time_alloc;
            obj.coeffs = obj.solve_minsnap();
        end

        function C = solve_minsnap(obj)
            % Solve QP for min-snap polynomials.
            % For brevity we use a closed-form 8th-order polynomial that
            % matches position, vel, accel, jerk = 0 at endpoints (rest-to-rest).
            N_seg = size(obj.waypoints, 1) - 1;
            C = zeros(4, 8, N_seg);

            for k = 1:N_seg
                T = obj.time_alloc(k);
                p0 = obj.waypoints(k,   :);
                p1 = obj.waypoints(k+1, :);
                for d = 1:4
                    % 8th-order polynomial: p(t) = sum_{j=0..7} c_j * t^j
                    % BCs: p(0)=p0, p'(0)=0, p''(0)=0, p'''(0)=0
                    %      p(T)=p1, p'(T)=0, p''(T)=0, p'''(T)=0
                    % Closed-form result yields a symmetric polynomial.
                    A = [1 0 0 0 0 0 0 0;
                         0 1 0 0 0 0 0 0;
                         0 0 2 0 0 0 0 0;
                         0 0 0 6 0 0 0 0;
                         1 T T^2 T^3 T^4 T^5 T^6 T^7;
                         0 1 2*T 3*T^2 4*T^3 5*T^4 6*T^5 7*T^6;
                         0 0 2 6*T 12*T^2 20*T^3 30*T^4 42*T^5;
                         0 0 0 6 24*T 60*T^2 120*T^3 210*T^4];
                    b = [p0(d); 0; 0; 0; p1(d); 0; 0; 0];
                    C(d, :, k) = (A \ b)';
                end
            end
        end

        function [pos, vel, acc, jerk, snap, psi_d, psi_dot] = eval(obj, t)
            % Evaluate trajectory at time t (0 <= t <= sum(time_alloc)).
            T_cumul = [0; cumsum(obj.time_alloc(:))];
            t = max(0, min(t, T_cumul(end)));
            k = find(t <= T_cumul, 1) - 1;
            if isempty(k) || k < 1, k = 1; end
            tau = t - T_cumul(k);

            powers   = tau .^ (0:7)';
            d1coef   = (0:7)';
            d2coef   = (0:7)' .* max(0, (-1:6)');
            d3coef   = (0:7)' .* max(0, (-1:6)') .* max(0, (-2:5)');
            d4coef   = (0:7)' .* max(0, (-1:6)') .* max(0, (-2:5)') .* max(0, (-3:4)');

            sigma   = zeros(4,1);
            sigmad  = zeros(4,1);
            sigmadd = zeros(4,1);
            sigmaddd  = zeros(4,1);
            sigmadddd = zeros(4,1);

            for d = 1:4
                c = squeeze(obj.coeffs(d, :, k))';
                sigma(d)   = c' * powers;
                sigmad(d)  = sum(d1coef .* c .* (tau.^max(0,(-1:6))'));
                sigmadd(d) = sum(d2coef .* c .* (tau.^max(0,(-2:5))'));
                sigmaddd(d) = sum(d3coef .* c .* (tau.^max(0,(-3:4))'));
                sigmadddd(d) = sum(d4coef .* c .* (tau.^max(0,(-4:3))'));
            end

            pos    = sigma(1:3);
            vel    = sigmad(1:3);
            acc    = sigmadd(1:3);
            jerk   = sigmaddd(1:3);
            snap   = sigmadddd(1:3);
            psi_d  = sigma(4);
            psi_dot = sigmad(4);
        end

        function [F_des_NED, qd, Wd] = recover_states(obj, t, mass, g)
            % Recover thrust vector + desired attitude from flat outputs.
            [~, ~, acc, jerk, ~, psi_d, ~] = obj.eval(t);
            F_des_NED = mass * (acc - [0;0;g]);   % g_NED has +Z down so subtract
            % qd via force_to_attitude logic (heading psi_d, thrust along F_des)
            % We replicate the mapping inline here for clarity.
            Fmag = norm(F_des_NED);
            if Fmag < 1e-3
                qd = quat_utils('fromEuler', 0, deg2rad(90), psi_d);
                Wd = zeros(3,1);
                return;
            end
            x_B = F_des_NED / Fmag;
            h = [cos(psi_d); sin(psi_d); 0];
            y_B = cross([0;0;-1], x_B);
            if norm(y_B) < 1e-3, y_B = cross(h, x_B); end
            y_B = y_B / norm(y_B);
            R = [x_B, y_B, cross(x_B, y_B)];
            qd = quat_utils('fromR', R);

            % Body angular velocity: derive from jerk projection (Mellinger 2011)
            % omega = R^T * (h_w x x_B) / |T|, where h_w = m * jerk
            h_w = mass * jerk;
            num = cross(h_w, x_B);
            Wd_world = num / max(Fmag, 1e-3);
            Wd = R' * Wd_world;
        end

        function T_total = total_time(obj)
            T_total = sum(obj.time_alloc);
        end
    end
end

File: evtol_python/evtol/trajectory/differential_flatness.py
"""Differential flatness trajectory in (x, y, z, psi) with min-snap polynomials.

8th-order rest-to-rest polynomials per segment, C^3 continuous.
Solved in NORMALIZED time u = tau/T_seg to avoid numerical ill-conditioning
when T is large (T^7 = 10^14 for T=100s would give cond > 10^16).
"""
from __future__ import annotations

import numpy as np


class DifferentialFlatness:
    """Multi-segment min-snap polynomial trajectory."""

    def __init__(self, waypoints: np.ndarray, time_alloc: np.ndarray):
        """
        Parameters
        ----------
        waypoints : (N+1, 4) array of [x, y, z, psi] waypoints
        time_alloc : (N,) array of segment durations
        """
        self.waypoints = np.asarray(waypoints, dtype=float)
        self.time_alloc = np.asarray(time_alloc, dtype=float).flatten()
        self.coeffs = self._solve_minsnap()
        self.T_cumul = np.concatenate([[0.0], np.cumsum(self.time_alloc)])

    def _solve_minsnap(self) -> np.ndarray:
        """Solve min-snap polynomial system in NORMALIZED time u = tau/T.

        Returns coeffs of shape (4, 8, N_seg).
        Polynomial in u: p(u) = sum_{j=0..7} c_j * u^j
        Derivatives in physical time: dp/dtau = (dp/du)/T, etc.
        """
        N_seg = self.waypoints.shape[0] - 1
        C = np.zeros((4, 8, N_seg))

        # Normalized A matrix (entries in {0, 1, 2, 6, 12, 20, 30, 42, 60, 120, 210})
        A = np.array([
            [1, 0, 0, 0,  0,  0,   0,   0],  # p(0)
            [0, 1, 0, 0,  0,  0,   0,   0],  # p'(0)
            [0, 0, 2, 0,  0,  0,   0,   0],  # p''(0)
            [0, 0, 0, 6,  0,  0,   0,   0],  # p'''(0)
            [1, 1, 1, 1,  1,  1,   1,   1],  # p(1)
            [0, 1, 2, 3,  4,  5,   6,   7],  # p'(1)
            [0, 0, 2, 6, 12, 20,  30,  42],  # p''(1)
            [0, 0, 0, 6, 24, 60, 120, 210],  # p'''(1)
        ], dtype=float)

        for k in range(N_seg):
            p0 = self.waypoints[k]
            p1 = self.waypoints[k + 1]
            for d in range(4):
                b = np.array([p0[d], 0, 0, 0, p1[d], 0, 0, 0])
                C[d, :, k] = np.linalg.solve(A, b)
        return C

    def total_time(self) -> float:
        return float(self.T_cumul[-1])

    def eval(self, t: float) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray, float, float]:
        """Evaluate trajectory at time t. Returns (pos, vel, acc, jerk, snap, psi, psi_dot)."""
        t = float(np.clip(t, 0.0, self.T_cumul[-1]))

        # Find segment
        k = max(0, int(np.searchsorted(self.T_cumul, t, side="right")) - 1)
        k = min(k, self.coeffs.shape[2] - 1)
        T_seg = self.time_alloc[k]
        tau = t - self.T_cumul[k]
        u = tau / T_seg

        # Normalized basis vectors
        U = np.array([u ** j for j in range(8)])
        Ud = np.concatenate([[0.0], [j * u ** (j - 1) for j in range(1, 8)]])
        Udd = np.concatenate([[0.0, 0.0],
                              [j * (j - 1) * u ** (j - 2) for j in range(2, 8)]])
        Uddd = np.concatenate([[0.0, 0.0, 0.0],
                               [j * (j - 1) * (j - 2) * u ** (j - 3) for j in range(3, 8)]])
        Udddd = np.concatenate([[0.0, 0.0, 0.0, 0.0],
                                [j * (j - 1) * (j - 2) * (j - 3) * u ** (j - 4) for j in range(4, 8)]])

        sigma = np.zeros(4)
        sigmad = np.zeros(4)
        sigmadd = np.zeros(4)
        sigmaddd = np.zeros(4)
        sigmadddd = np.zeros(4)

        for d in range(4):
            c = self.coeffs[d, :, k]
            sigma[d] = c @ U
            sigmad[d] = (c @ Ud) / T_seg
            sigmadd[d] = (c @ Udd) / T_seg ** 2
            sigmaddd[d] = (c @ Uddd) / T_seg ** 3
            sigmadddd[d] = (c @ Udddd) / T_seg ** 4

        return sigma[:3], sigmad[:3], sigmadd[:3], sigmaddd[:3], sigmadddd[:3], float(sigma[3]), float(sigmad[3])
